server/src/services/documentSymbol/SymbolVisitor.ts

```diff
@@ -0,0 +1,68 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+import { SymbolKind } from "vscode-languageserver-types";
+import _ from "lodash";
+import { RuleKind, TokenKind } from "@nomicfoundation/slang/kinds";
+import { RuleNode, TokenNode } from "@nomicfoundation/slang/cst";
+import { TextDocument } from "vscode-languageserver-textdocument";
+import { Cursor } from "@nomicfoundation/slang/cursor";
+import { slangToVSCodeRange } from "../../parser/slangHelpers";
+import { SymbolTreeBuilder } from "./SymbolTreeBuilder";
+
+export abstract class SymbolVisitor {
+  public abstract ruleKind: RuleKind;
+  public abstract symbolKind: SymbolKind;
+  public abstract nameTokenKind: TokenKind;
+
+  constructor(
+    public document: TextDocument,
+    public symbolBuilder: SymbolTreeBuilder
+  ) {}
+
+  public onRuleNode(node: RuleNode, cursor: Cursor): void {
```

Comment:
Not sure I understand why `SymbolVisitor` is an `abstract class`, given that all its children are just data pairs. Are we planning on having custom logic per visitor? Otherwise, I wonder if this can be simplified into a simple map/dictionary of `ruleKind -> (symbolKind, nameTokenKind)`.

Reply:
Previously there was custom logic for some of the child classes, but it was changed a lot when it was reworked to use the Cursor api. I'll refactor it

Reply:
Actually I think it's ok to leave this as an object hierarchy. It gives us flexibility in case we want to change the abstraction, push some methods down to subclasses, etc. If I refactor this into a dictionary and then need to leverage those oop features I'd have to convert it again to an oop model.

Reply:
Alright. Let's keep it for now. Thanks!